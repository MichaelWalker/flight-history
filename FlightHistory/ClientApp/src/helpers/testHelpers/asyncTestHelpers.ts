import { act } from "@testing-library/react-hooks";

export async function runActionAndWaitForAsyncActionsToComplete(
    action: () => Promise<unknown> | void,
): Promise<void> {
    await act(async () => {
        await action();
    });
}

export async function quickResolve<T>(callback: () => T): Promise<T> {
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve(callback());
        }, 0);
    });
}

export async function quickReject<T>(callback: () => T): Promise<T> {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            reject(callback());
        }, 0);
    });
}

export class WrappedPromise<T> {
    private readonly promise: Promise<T>;
    private resolved = false;
    private result: T | undefined;

    constructor(promise: Promise<T>) {
        this.promise = promise;

        promise.then((result) => {
            this.resolved = true;
            this.result = result;
        });
    }

    isResolved(): boolean {
        return this.resolved;
    }
}
