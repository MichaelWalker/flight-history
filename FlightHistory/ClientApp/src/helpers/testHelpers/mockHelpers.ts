import * as globals from "../globals";

export function mockUseSampleData(useSampleData: boolean) {
    jest.spyOn(globals, "useSampleData").mockReturnValue(useSampleData);
}

export function mockUseConsoleLogging(useConsoleLogging: boolean) {
    jest.spyOn(globals, "useConsoleLogging").mockReturnValue(useConsoleLogging);
}
