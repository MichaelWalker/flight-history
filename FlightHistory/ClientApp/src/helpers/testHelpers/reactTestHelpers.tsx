import type { ReactElement } from "react";
import React from "react";
import type { RenderResult } from "@testing-library/react";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export function renderWithRouter(component: ReactElement): RenderResult {
    return render(<MemoryRouter>{component}</MemoryRouter>);
}

export async function waitForTableToHaveResults(): Promise<void> {
    await waitFor(() => expect(screen.getAllByRole("row").length).toBeGreaterThan(1));
}
