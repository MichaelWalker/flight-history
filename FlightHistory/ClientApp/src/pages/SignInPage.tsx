import React, {FunctionComponent, useCallback, useState} from "react";
import {EmailInput, PasswordInput} from "../components/forms/Input";
import {ApiClient} from "../api/apiClient";
import styles from "./SignInPage.module.scss";
import {Form} from "../components/forms/Form";
import {ApiError} from "../api/apiError"; 

export const SignInPage: FunctionComponent = () => {
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    
    const validateCallback = useCallback(() => {
        return !(password === "" || email === "");
        
    }, [password, email]);
    
    function errorMessage(error: ApiError) {
        if (error.status === 401) {
            return "Incorrect email or password";
        }
        return "Sorry - something went wrong. Please try again."
    }
        
    function signIn() {
        return ApiClient.signIn(email, password);
    }
    
    return (
        <main className={styles.main}>
            <section className={styles.section}>
                <h1 className={styles.title}>Sign In</h1>
                <Form buttonText={'Sign In'} apiRequest={signIn} errorMessage={errorMessage} validateCallback={validateCallback}>
                    <EmailInput value={email} updateValue={setEmail} required={true}>Email</EmailInput>
                    <PasswordInput value={password} updateValue={setPassword} required={true}>Password</PasswordInput>
                </Form>
            </section>
        </main>
    );
};