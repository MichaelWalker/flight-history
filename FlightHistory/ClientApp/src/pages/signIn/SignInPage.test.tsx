import "isomorphic-fetch";
import React from "react";
import { renderWithRouter } from "../../helpers/testHelpers/reactTestHelpers";
import { SignInPage } from "./SignInPage";

describe("SignIn Page", () => {
    it("matches snapshot", async () => {
        const { container } = renderWithRouter(<SignInPage />);
        expect(container).toMatchSnapshot();
    });
});
