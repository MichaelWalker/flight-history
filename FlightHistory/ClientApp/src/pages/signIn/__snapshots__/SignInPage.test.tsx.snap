// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Page matches snapshot 1`] = `
<div>
  <main
    class="main"
  >
    <section
      class="section"
    >
      <h1
        class="title"
        data-testid="page-title"
      >
        Sign In
      </h1>
      <form>
        <label
          class="css-oeotp9"
        >
          <span
            class="css-ar6pse"
          >
            Email
          </span>
          <input
            class="css-1ielmxe"
            required=""
            type="email"
            value=""
          />
        </label>
        <label
          class="css-oeotp9"
        >
          <span
            class="css-ar6pse"
          >
            Password
          </span>
          <input
            class="css-1ielmxe"
            required=""
            type="password"
            value=""
          />
        </label>
        <button
          class="button"
          disabled=""
          type="submit"
        >
          Sign In
        </button>
        <div
          class="css-1btb9lf"
        />
      </form>
    </section>
  </main>
</div>
`;
